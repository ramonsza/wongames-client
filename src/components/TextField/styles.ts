import styled, { DefaultTheme, css } from 'styled-components'

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};
    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`

type IconPositionProps = {
  $iconPosition: 'left' | 'right'
}

export const Icon = styled.div<IconPositionProps>`
  ${({ theme, $iconPosition }) => css`
    display: flex;
    justify-content: center;
    align-items: center;
    width: 2.2rem;
    color: ${theme.colors.gray};
    & > svg {
      width: 100%;
    }

    order: ${$iconPosition === 'left' ? 0 : 1};
  `}
`

export const Input = styled.input<IconPositionProps>`
  ${({ theme, $iconPosition }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
    padding-${$iconPosition}: ${theme.spacings.xxsmall};
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
  `}
`

const wrapperModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${Input},
    ${Icon} {
      cursor: not-allowed;
      color: ${theme.colors.gray};
      &::placeholder {
        color: currentColor;
      }
    }
  `
}

type WrapperProps = {
  $disable: boolean
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, $disable }) => css`
    ${$disable && wrapperModifiers.disabled(theme)}
  `}
`
