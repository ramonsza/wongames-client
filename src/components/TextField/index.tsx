import { useState, InputHTMLAttributes } from 'react'

import * as S from './styles'

export type TextFieldProps = {
  onInput?: (value: string) => void
  label?: string
  labelFor?: string
  initialValue?: string
  icon?: React.ReactNode
  iconPosition?: 'left' | 'right'
  disabled?: boolean
} & InputHTMLAttributes<HTMLInputElement>

const TextField = ({
  label,
  labelFor = '',
  initialValue = '',
  iconPosition = 'left',
  disabled = false,
  icon,
  onInput,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue)

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value
    setValue(newValue)

    !!onInput && onInput(newValue)
  }

  return (
    <S.Wrapper $disable={disabled}>
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}
      <S.InputWrapper>
        {!!icon && <S.Icon $iconPosition={iconPosition}>{icon}</S.Icon>}
        <S.Input
          $iconPosition={iconPosition}
          type="text"
          onChange={onChange}
          value={value}
          disabled={disabled}
          {...props}
        />
      </S.InputWrapper>
    </S.Wrapper>
  )
}

export default TextField
