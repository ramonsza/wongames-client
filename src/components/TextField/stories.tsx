import { EnvelopeSimple as MailIcon } from '@phosphor-icons/react'
import { Meta, StoryObj } from '@storybook/react'

import TextField from '.'

export default {
  title: 'Form/TextField',
  component: TextField,
  args: {
    label: 'E-mail',
    labelFor: 'Email',
    id: 'Email',
    initialValue: '',
    placeholder: 'john.cage@gmail.com'
  },
  argTypes: {
    onInput: { action: 'changed' },
    icon: {
      control: false
    }
  }
} as Meta

type Story = StoryObj<typeof TextField>

export const Default: Story = {
  render: (args) => (
    <div style={{ maxWidth: 300, padding: 15 }}>
      <TextField {...args} />
    </div>
  ),
  args: {
    icon: <MailIcon size={22}></MailIcon>
  }
}

export const WithError: Story = {
  render: (args) => (
    <div style={{ maxWidth: 300, padding: 15 }}>
      <TextField {...args} />
    </div>
  ),
  args: {
    icon: <MailIcon size={22}></MailIcon>,
    error: 'Ops... something is wrong'
  }
}
