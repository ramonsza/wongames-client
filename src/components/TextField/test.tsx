import { screen, waitFor } from '@testing-library/react'
import { EnvelopeSimple as MailIcon } from '@phosphor-icons/react'
import TextField from '.'
import { renderWithTheme } from '@/utils/tests/helpers'
import userEvent from '@testing-library/user-event'

describe('<TextField />', () => {
  it('Renders with Label', () => {
    renderWithTheme(<TextField label="Label" labelFor="Field" id="Field" />)

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('Renders without Label', () => {
    renderWithTheme(<TextField />)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('Renders with placeholder', () => {
    renderWithTheme(<TextField placeholder="hey you" />)

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
  })

  it('Changes its value when typing', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <TextField
        onInput={onInput}
        label="TextField"
        labelFor="TextField"
        id="TextField"
      />
    )

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInput).toHaveBeenCalledTimes(text.length)
    })
    expect(onInput).toHaveBeenCalledWith(text)
  })

  it('Is accessible by tab', async () => {
    renderWithTheme(
      <TextField label="TextField" labelFor="TextField" id="TextField" />
    )

    const input = screen.getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    await userEvent.tab()
    expect(input).toHaveFocus()
  })

  it('Renders with Icon', () => {
    renderWithTheme(<TextField icon={<MailIcon data-testid="icon" />} />)

    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('Renders with Icon on the right side', () => {
    renderWithTheme(
      <TextField icon={<MailIcon data-testid="icon" />} iconPosition="right" />
    )

    expect(screen.getByTestId('icon').parentElement).toHaveStyle({ order: 1 })
  })

  it('Does not changes its value when disabled', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <TextField
        onInput={onInput}
        label="TextField"
        labelFor="TextField"
        id="TextField"
        disabled
      />
    )

    const input = screen.getByRole('textbox')
    expect(input).toBeDisabled()

    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).not.toHaveValue(text)
    })
    expect(onInput).not.toHaveBeenCalled()
  })
})
