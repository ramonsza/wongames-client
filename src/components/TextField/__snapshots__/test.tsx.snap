// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Renders with error 1`] = `
.c4 {
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c6 {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 2.2rem;
  color: #8F8F8F;
  order: 0;
}

.c6>svg {
  width: 100%;
}

.c7 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  padding-left: 0.8rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c2 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c0 .c3 {
  border-color: #FF6347;
}

.c0 .c1,
.c0 .c5 {
  color: #FF6347;
}

.c8 {
  color: #FF6347;
  font-size: 1.2rem;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="TextField"
  >
    TextField
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="c5 c6"
    >
      <svg
        data-testid="icon"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224,48H32a8,8,0,0,0-8,8V192a16,16,0,0,0,16,16H216a16,16,0,0,0,16-16V56A8,8,0,0,0,224,48ZM203.43,64,128,133.15,52.57,64ZM216,192H40V74.19l82.59,75.71a8,8,0,0,0,10.82,0L216,74.19V192Z"
        />
      </svg>
    </div>
    <input
      class="c7"
      id="TextField"
      name="TextField"
      type="text"
      value=""
    />
  </div>
  <p
    class="c8"
  >
    Error message
  </p>
</div>
`;
